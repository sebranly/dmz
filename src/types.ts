// TODO: remove
export enum Template {
  Typescript = 'typescript',
  Javascript = 'javascript'
}

export type TemplateLanguage = {
  language: Template;
};
