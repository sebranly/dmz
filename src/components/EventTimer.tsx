import classnames from 'classnames';
import * as React from 'react';
import { APITime, TimeUnit } from '../types';
import { getDateTime } from '../utils';
import { convertSecondsToTimeValue } from '../utils/convert';
import { displayWithTwoDigits, getStatusColor, getStatusVerb, getTimeUnitAbbreviation } from '../utils/display';
import { getTimerClasses } from '../utils/tailwind';

export interface EventTimerProps {
  className?: string;
  currentTimestamp: number;
  time: APITime;
}

const EventTimer: React.FC<EventTimerProps> = (props) => {
  const { className, currentTimestamp, time } = props;
  const { name, time: eventTime, status } = time;

  const remainingSeconds = eventTime - currentTimestamp;
  const isPast = remainingSeconds <= 0;
  const statusVerb = getStatusVerb(status);

  const subtitle = isPast ? `It ${statusVerb}ed already` : `It ${statusVerb}es in`;
  const color = getStatusColor(status);

  const classnamesColor = `text-${color}-500`;
  const classnamesSubtitle = 'font-bold my-1 text-lg';
  const classnamesTitle = classnames(classnamesSubtitle, classnamesColor);
  const classnamesTime = classnames('font-bold pr-1', classnamesColor);

  const {
    [TimeUnit.Day]: days,
    [TimeUnit.Hour]: hours,
    [TimeUnit.Minute]: minutes,
    [TimeUnit.Second]: seconds
  } = convertSecondsToTimeValue(remainingSeconds);

  const isFixedDays = days === 0;
  const isFixedHours = isFixedDays && hours === 0;
  const isFixedMinutes = isFixedHours && minutes === 0;
  const isFixedSeconds = isFixedMinutes && seconds === 0;

  const classnamesComponent = getTimerClasses(color, className);

  const items = [
    { value: days, label: TimeUnit.Day },
    { value: hours, label: TimeUnit.Hour },
    { value: minutes, label: TimeUnit.Minute },
    { value: seconds, label: TimeUnit.Second }
  ].map((element: { value: number; label: TimeUnit }) => {
    const { value, label: l } = element;
    const isFixed =
      (l === TimeUnit.Day && isFixedDays) ||
      (l === TimeUnit.Hour && isFixedHours) ||
      (l === TimeUnit.Minute && isFixedMinutes) ||
      (l === TimeUnit.Second && isFixedSeconds);

    const classnamesValue = classnames('font-bold text-4xl', { [classnamesColor]: isFixed });

    return (
      <li className="inline-block mx-2.5" key={l}>
        <span className={classnamesValue}>{displayWithTwoDigits(value)}</span>
        <div className="text-sm">{getTimeUnitAbbreviation(l)}</div>
      </li>
    );
  });

  return (
    <div className={classnamesComponent}>
      <div className={classnamesTitle}>{name}</div>
      <div className={classnamesSubtitle}>{subtitle}</div>
      <ul className="timer-card flex justify-center">{items}</ul>
      {isPast ? (
        <div className="text-xs sm:text-sm">The website will be updated in the next few days</div>
      ) : (
        <div className="text-xs sm:text-sm">
          <div className="flex text-left pl-2.5">
            <div className="grow">Release Date:</div> <div className={classnamesTime}>{getDateTime(eventTime)}</div>
          </div>
        </div>
      )}
    </div>
  );
};

export { EventTimer };
