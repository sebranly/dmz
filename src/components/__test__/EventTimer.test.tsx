import * as React from 'react';
import { render } from '@testing-library/react';
import { EventTimer } from '../EventTimer';
import { APITimer, TimerType } from '../../types';

test('EventTimer', () => {
  const timer: APITimer = {
    type: TimerType.Event,
    title: 'Season 03',
    showPostEvent: true,
    subtitle: 'It launches in',
    subtitlePostEvent: 'It launched already',
    data: [{
      color: 'red',
      description: 'Release Date:',
      time: 1678125600
    }]
  };

  const createComponent = (currentTimestamp: number) => (
    <EventTimer currentTimestamp={currentTimestamp} timer={timer} />
  );

  const { container: containerPostEvent } = render(createComponent(1678125601));
  expect(containerPostEvent.childNodes).toMatchSnapshot();

  const { container: containerEventIncoming } = render(createComponent(1677952801));
  expect(containerEventIncoming.childNodes).toMatchSnapshot();
});
