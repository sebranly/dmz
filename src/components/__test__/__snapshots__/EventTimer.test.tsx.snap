// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventTimer 1`] = `NodeList []`;

exports[`EventTimer 2`] = `
NodeList [
  <div
    class="bg-neutral-800 border-solid border-2 rounded-lg border-red-500 m-2.5 p-2.5 relative w-72 sm:w-80"
  >
    <div
      class="font-bold my-1 text-lg text-red-500"
    >
      Season 03
    </div>
    <div
      class="font-bold my-1 text-lg"
    >
      It launches in
    </div>
    <ul
      class="timer-card flex justify-center"
    >
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          01
        </span>
        <div
          class="text-sm"
        >
          days
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          23
        </span>
        <div
          class="text-sm"
        >
          hrs
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          59
        </span>
        <div
          class="text-sm"
        >
          min
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          59
        </span>
        <div
          class="text-sm"
        >
          sec
        </div>
      </li>
    </ul>
    <div
      class="text-xs sm:text-sm"
    >
      <div
        class="flex text-left pl-2.5"
      >
        <div
          class="grow"
        >
          Release Date:
        </div>
         
        <div
          class="font-bold pr-1 text-red-500"
        >
          Mon 3/6 07:00â€¯PM
        </div>
      </div>
    </div>
  </div>,
]
`;
